src/System.IO.Compression/tests/DeflateStreamTests.cs

```diff
@@ -15,6 +15,95 @@ public class DeflateStreamTests
+                {
+                    // Unblock Async operations
+                    readStream.manualResetEvent.Set();
+                    // The original ReadAsync should be able to complete
+                    Assert.True(task.Wait(100 * 500));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingFlushAsync()
+        {
+            byte[] buffer = null;
+            string testFilePath = gzTestFile("GZTestDocument.pdf");
+            using (var origStream = await LocalMemoryStream.readAppFileAsync(testFilePath))
+            {
+                buffer = origStream.ToArray();
+            }
+
+            using (var writeStream = new ManualSyncMemoryStream(false))
+            using (var zip = new DeflateStream(writeStream, CompressionMode.Compress))
+            {
+                Task task = null;
+                try
+                {
+                    writeStream.manualResetEvent.Set();
+                    await zip.WriteAsync(buffer, 0, buffer.Length);
+                    writeStream.manualResetEvent.Reset();
+                    writeStream.WriteHit = false;
+                    task = zip.FlushAsync();
+                    Assert.True(writeStream.WriteHit);
+                    Assert.Throws<InvalidOperationException>(() => { zip.FlushAsync(); }); // "overlapping flushes"
+                }
+                finally
+                {
+                    // Unblock Async operations
+                    writeStream.manualResetEvent.Set();
+                    // The original WriteAsync should be able to complete
+                    Assert.True(task.Wait(100 * 500));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingWriteAsync()
+        {
+            byte[] buffer = null;
+            string testFilePath = gzTestFile("GZTestDocument.pdf");
+            using (var origStream = await LocalMemoryStream.readAppFileAsync(testFilePath))
+            {
+                buffer = origStream.ToArray();
+            }
+
+            using (var writeStream = new ManualSyncMemoryStream(false))
+            using (var zip = new DeflateStream(writeStream, CompressionMode.Compress))
+            {
+                Task task = null;
+                try
+                {
+                    task = zip.WriteAsync(buffer, 0, buffer.Length);    // write needs to be bigger than the internal write buffer
+                    Assert.True(writeStream.WriteHit);
+                    Assert.Throws<InvalidOperationException>(() => { zip.WriteAsync(buffer, 32, 32); }); // "overlapping write"
+                }
+                finally
+                {
+                    // Unblock Async operations
+                    writeStream.manualResetEvent.Set();
+                    // The original WriteAsync should be able to complete
+                    Assert.True(task.Wait(100 * 500));
+                }
+            }
+        }
```

Comment:
Sure, we can do that. We're only missing two scenarios: Flush during write, and Flush during read. Write during read and read during write will fail because of the CompressionMode being incorrect before they fail from the concurrent async operations check.